### PR TITLE
fix: request.clone() actually clones the request body.

### LINKED ISSUE #1381: `Request.clone()` does not clone or support streaming body
### What version of Bun is running?

0.2.1 (canary)

### What platform is your computer?

Darwin 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 20:57:23 PDT 2022; root:xnu-8020.121.3~2/RELEASE_ARM64_T8110 arm64 arm

### What steps can reproduce the bug?

```ts
export default {
  async fetch(request) {
    console.log(".clone().json()", await request.clone().json());
    console.log(".json()", await request.json());
    return new Response("Ok");
  }
}
```

`curl -d '{}' http://localhost:3000`

### How often does it reproduce? Is there a required condition?

Always. Same with `.text()`.

### What is the expected behavior?

_No response_

### What do you see instead?

Terminal output:
```bash
.clone().json() # body is missing
.json() {} # works fine
```

### Additional information

_No response_

### LINKED ISSUE #8515: `Request.clone()` does not clone body
### What version of Bun is running?

1.0.25+a8ff7be64

### What platform is your computer?

macOS

### What steps can reproduce the bug?

clone an incoming request

### What is the expected behavior?

the body is cloned

### What do you see instead?

blank body

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

This makes request.clone() call .body.tee(), and then re-assigns request.body to the teed ReadableStream. 

Fixes #8515
Fixes #1381

### How did you verify your code works?

This needs tests.

### CODE REVIEW

File: src/bun.js/bindings/ZigGlobalObject.cpp [hunk 1]
```diff
@@ -2284,6 +2284,18 @@ extern "C" void ReadableStream__cancel(JSC__JSValue possibleReadableStream, Zig:
     ReadableStream::cancel(*globalObject, readableStream, exception);
 }
 
+extern "C" void ReadableStream__tee(JSC__JSValue possibleReadableStream, Zig::GlobalObject* globalObject, JSC__JSValue ptr[2])
+{
+    auto* readableStream = jsDynamicCast<JSReadableStream*>(JSC::JSValue::decode(possibleReadableStream));
+    if (UNLIKELY(!readableStream))
```
Comment (necmettindev) [inline src/bun.js/bindings/ZigGlobalObject.cpp:2290]: just wondering, is there a specific reason why we don't use `LIKELY(readableStream)?`

File: src/bun.js/bindings/webcore/ReadableStream.cpp [hunk 2]
```diff
@@ -125,24 +126,25 @@ void ReadableStream::pipeTo(ReadableStreamSink& sink)
...
     if (!returnedValue)
         return {};
 
-    auto results = Detail::SequenceConverter<IDLInterface<ReadableStream>>::convert(lexicalGlobalObject, *returnedValue);
+    JSValue array = returnedValue.value();
+    auto* arrayObject = jsCast<JSArray*>(array.getObject());
+    auto first = arrayObject->getDirectIndex(&lexicalGlobalObject, 0);
+    auto second = arrayObject->getDirectIndex(&lexicalGlobalObject, 1);
 
-    ASSERT(results.size() == 2);
```
Comment (paperclover) [inline src/bun.js/bindings/webcore/ReadableStream.cpp:144]: why was this removed

File: src/bun.js/bindings/ZigGlobalObject.cpp (hunk 1 above)
Comment (paperclover) [reply src/bun.js/bindings/ZigGlobalObject.cpp:2290]: this condition is likely to be true: `!readableStream`

This path only happens if you call this with a |this| value that is not JSReadableStream.

**Comment by paperclover** [comment]
cannot merge because of failing test case. feature works for simpler cases but this may not ship with this memory corruption

image

**Comment by KilianB** [comment]
Would it be worth to also support the copy constructor of `Request` in the same breath? https://developer.mozilla.org/en-US/docs/Web/API/Request/Request Open Bun issue: https://github.com/oven-sh/bun/issues/4728. This would allow to clone the request while adjusting individual fields like the url parameter, fix a bug in elysia (e.g. https://github.com/elysiajs/elysia/issues/332) where mounted routes are not able to access the body.

**Comment by beorn** [comment]
Until this is implemented, perhaps just have `req.clone()` core dump / panic instead of leading to lots of hard-to-debug downstream errors?

**Comment by billywhizz** [comment]
we investigated [an issue](https://github.com/oven-sh/bun/issues/2644) last week with formdata() not working as expected and it seems to be related to the issue addressed by this PR. when instantiating a new Request object and passing an existing request as the first argument to the constructor, we do not clone the body as required by the spec.

This is a list of known issues which are caused by this missing request body clone functionality:

## related issues
- https://github.com/oven-sh/bun/issues/2644
- https://github.com/oven-sh/bun/issues/10789
- https://github.com/oven-sh/bun/issues/4161
- https://github.com/oven-sh/bun/issues/4728

**Comment by Jarred-Sumner** [comment]
Superseded by https://github.com/oven-sh/bun/pull/12473
